<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>irg Vignette</title>

<script>// Hide empty <a> tag within highlighted CodeBlock for screen reader accessibility (see https://github.com/jgm/pandoc/issues/6352#issuecomment-626106786) -->
// v0.0.1
// Written by JooYoung Seo (jooyoung@psu.edu) and Atsushi Yasumoto on June 1st, 2020.

document.addEventListener('DOMContentLoaded', function() {
  const codeList = document.getElementsByClassName("sourceCode");
  for (var i = 0; i < codeList.length; i++) {
    var linkList = codeList[i].getElementsByTagName('a');
    for (var j = 0; j < linkList.length; j++) {
      if (linkList[j].innerHTML === "") {
        linkList[j].setAttribute('aria-hidden', 'true');
      }
    }
  }
});
</script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore"><code>irg</code> Vignette</h1>



<p>The <code>irg</code> package is based on the procedure proposed in <span class="citation">Heerah et al. (2021)</span>.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw">library</span>(irg)</span></code></pre></div>
<p>The functions made available allow to perform Granger-Causal testing for signals that are sampled at the same but irregularly spaced time points. Before using the functions in the package, the signals need to be pre-processed such that they are detrended (mean-stationary) and standardized (normalized). Below we show some examples of possible ways of pre-processing the signals as well as how to use the main functions of the package.</p>
<div id="data-pre-processing" class="section level2">
<h2>Data Pre-Processing</h2>
<p>In this section we use data that was collected during an experiment that measured the time-evolved transcriptome of the three replicates of Arabidopsis roots and shoots. The data <code>raw_data</code> contains a matrix reporting a gene expression count for root and shoot collected at the following time points: 0, 5, 10, 15, 20, 30, 45, 60, 90 and 120 (minutes).</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="kw">data</span>(raw_data)</span>
<span id="cb2-2"><a href="#cb2-2"></a><span class="kw">head</span>(raw_data)</span></code></pre></div>
<pre><code>##      root shoot time
## [1,] 4965   674    0
## [2,] 4776   714    0
## [3,] 4806   796    0
## [4,] 4890   540    5
## [5,] 3896   722    5
## [6,] 4441   578    5</code></pre>
<p>We decide to use the following simple functions to pre-process the data in this matrix (but other pre-processing techniques can be chosen).</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a><span class="kw">library</span>(mgcv)</span>
<span id="cb4-2"><a href="#cb4-2"></a></span>
<span id="cb4-3"><a href="#cb4-3"></a><span class="co"># Average signals at matching time points and then standardize</span></span>
<span id="cb4-4"><a href="#cb4-4"></a>avg &lt;-<span class="st"> </span><span class="cf">function</span>(signals) {</span>
<span id="cb4-5"><a href="#cb4-5"></a></span>
<span id="cb4-6"><a href="#cb4-6"></a>  signal &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="ot">NA</span>, <span class="kw">length</span>(signals)<span class="op">/</span><span class="dv">3</span>)</span>
<span id="cb4-7"><a href="#cb4-7"></a></span>
<span id="cb4-8"><a href="#cb4-8"></a>  <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>(<span class="kw">length</span>(signals)<span class="op">/</span><span class="dv">3</span>)) {</span>
<span id="cb4-9"><a href="#cb4-9"></a></span>
<span id="cb4-10"><a href="#cb4-10"></a>    signal[i] &lt;-<span class="st"> </span><span class="kw">mean</span>(signals[(<span class="dv">3</span><span class="op">*</span>(i<span class="dv">-1</span>) <span class="op">+</span><span class="st"> </span><span class="dv">1</span>) <span class="op">:</span><span class="st"> </span>(<span class="dv">3</span><span class="op">*</span>i)])</span>
<span id="cb4-11"><a href="#cb4-11"></a></span>
<span id="cb4-12"><a href="#cb4-12"></a>  }</span>
<span id="cb4-13"><a href="#cb4-13"></a></span>
<span id="cb4-14"><a href="#cb4-14"></a>  signal_std &lt;-<span class="st"> </span>(signal <span class="op">-</span><span class="st"> </span><span class="kw">mean</span>(signal))<span class="op">/</span><span class="kw">sd</span>(signal)</span>
<span id="cb4-15"><a href="#cb4-15"></a>  </span>
<span id="cb4-16"><a href="#cb4-16"></a>  <span class="kw">return</span>(signal_std)</span>
<span id="cb4-17"><a href="#cb4-17"></a></span>
<span id="cb4-18"><a href="#cb4-18"></a>}</span>
<span id="cb4-19"><a href="#cb4-19"></a></span>
<span id="cb4-20"><a href="#cb4-20"></a><span class="co"># Pre-processing function that uses GAM to detrend the signals and then uses previous function to aggregate and normalize</span></span>
<span id="cb4-21"><a href="#cb4-21"></a>pre_process &lt;-<span class="st"> </span><span class="cf">function</span>(dat_mat) {</span>
<span id="cb4-22"><a href="#cb4-22"></a></span>
<span id="cb4-23"><a href="#cb4-23"></a>  root &lt;-<span class="st"> </span>dat_mat[,<span class="dv">1</span>]</span>
<span id="cb4-24"><a href="#cb4-24"></a>  shoot &lt;-<span class="st"> </span>dat_mat[,<span class="dv">2</span>]</span>
<span id="cb4-25"><a href="#cb4-25"></a>  time &lt;-<span class="st"> </span>dat_mat[,<span class="dv">3</span>]</span>
<span id="cb4-26"><a href="#cb4-26"></a></span>
<span id="cb4-27"><a href="#cb4-27"></a>  mod_root &lt;-<span class="st"> </span><span class="kw">gam</span>(root <span class="op">~</span><span class="st"> </span><span class="kw">s</span>(time))</span>
<span id="cb4-28"><a href="#cb4-28"></a>  mod_shoot &lt;-<span class="st"> </span><span class="kw">gam</span>(shoot <span class="op">~</span><span class="st"> </span><span class="kw">s</span>(time))</span>
<span id="cb4-29"><a href="#cb4-29"></a></span>
<span id="cb4-30"><a href="#cb4-30"></a>  resid_root &lt;-<span class="st"> </span><span class="kw">avg</span>(<span class="kw">resid</span>(mod_root))</span>
<span id="cb4-31"><a href="#cb4-31"></a>  resid_shoot &lt;-<span class="st"> </span><span class="kw">avg</span>(<span class="kw">resid</span>(mod_shoot))</span>
<span id="cb4-32"><a href="#cb4-32"></a></span>
<span id="cb4-33"><a href="#cb4-33"></a>  out &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">root =</span> resid_root, <span class="dt">shoot =</span> resid_shoot)</span>
<span id="cb4-34"><a href="#cb4-34"></a></span>
<span id="cb4-35"><a href="#cb4-35"></a>  <span class="kw">return</span>(out)</span>
<span id="cb4-36"><a href="#cb4-36"></a></span>
<span id="cb4-37"><a href="#cb4-37"></a>}</span></code></pre></div>
<p>If we apply the `<code>pre_process</code> function to the raw signals we obtain the following:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a>signals &lt;-<span class="st"> </span><span class="kw">pre_process</span>(raw_data)</span>
<span id="cb5-2"><a href="#cb5-2"></a>signals</span></code></pre></div>
<pre><code>## $root
##  [1]  1.71628640 -0.04956446 -1.59010074 -0.59163686 -0.79322148  0.73618577
##  [7]  0.14033187  1.08903416 -0.86743386  0.21011919
## 
## $shoot
##  [1]  1.89019661 -0.71701506 -1.10224276 -0.54846938 -1.11158441  1.36759941
##  [7]  0.39456993  0.05497426 -0.34491931  0.11689071</code></pre>
<p>The result of the pre-processing is saved in the <code>signals</code> data made available with the <code>irg</code> package.</p>
</div>
<div id="granger-causal-testing" class="section level2">
<h2>Granger-Causal Testing</h2>
<p>Once the signals have been pre-processed, it is possible to make use of the <code>irg</code> package functions and for this we make use of the pre-processed signals made available in the package: <code>arabidopsis</code>.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="kw">data</span>(arabidopsis)</span></code></pre></div>
<p>This is a list containing two dataframes, for root (<code>$roots</code>) and shoot (<code>$shoots</code>) signals respectively, containing the gene expressions collected over the following time points: 0, 5, 10, 15, 20, 30, 45, 60, 90, 120 (in minutes). It can be noticed how the time intervals are irregular between measurements and therefore the <code>irg</code> package makes available a Granger-Causal testing framework for these specific settings.</p>
<p>The main function of the package is the <code>granger_test</code> function. The null hypothesis for all tests that this function implements is the following:</p>
<p>“<span class="math inline">\(H_0\)</span>: The two signals have no Granger-Causal impact on each other”.</p>
<p>The goal of the function is to understand whether this null hypothesis can be rejected in favour of a specific alternative hypothesis, an example of which is given by:</p>
<p>“<span class="math inline">\(H_A\)</span>: The first signal (root) has a Granger-Causal impact on the second signal (shoot)”.</p>
<p>In order to perform this test of hypotheses, <span class="citation">Heerah et al. (2021)</span> put forward a bivariate autoregressive model for irregularly sampled signals mainly characterized by the following parameter vector (other parameters not in this vector are implicitly defined in the function):</p>
<p><span class="math display">\[\theta := \big[\phi_{root},\, \psi_{root},\, \gamma_{root},\, \sigma_{root}^2,\, \phi_{shoot},\, \psi_{shoot},\, \gamma_{shoot},\, \sigma_{shoot}^2\big].\]</span> The subscripts “root” and “shoot” underline how these parameters contribute to the behaviour over time of either the first signal (root) or second signal (shoot). While the <span class="math inline">\(\phi\)</span> and <span class="math inline">\(\sigma^2\)</span> parameters are specific to the dependence of a signal only on its own past (and represent the range of dependence on the past and the residual variance respectively), the main parameters of interest for the mentioned Granger-Causal testing framework are the <span class="math inline">\(\psi\)</span> and <span class="math inline">\(\gamma\)</span> parameters which represent the “intensity” and “time” of maximal (Granger-Causal) impact of one signal on the other. More specifically, <span class="math inline">\(\psi \in (-1, 1)\)</span> can be interpreted similarly to Pearson’s correlation coefficient: the closer the value is to 1 (in absolute value) the greater is the impact of one signal on the other, while the sign represents the direction of the impact where, for example, a negative value implies that the impacted signal grows as the “causing” signal descreases (and viceversa). The <span class="math inline">\(\gamma \in \mathbb{R}^+\)</span> parameter instead represents the distance in time at which the impact of one signal on the other is maximal (under the assumption that the impact increases and then decreases monotonically over time). So, for example, <span class="math inline">\(\psi_{root}\)</span> represents the intensity of the impact of the shoot (second signal) on the root (first signal) while <span class="math inline">\(\gamma_{root}\)</span> represents the time required for the impact of the shoot on the root to be maximal.</p>
<p>Estimating the parameters of these models via maximum-likelihood (under a Gaussian assumption), it is then possible to perform a Likelihood Ratio Test (LRT) to understand if the model under the alternative hypothesis significantly improves the fit of the signals. Given the short length of signals usually measured in biological experiments, the LRT test is performed using parametric bootstrap to approximate the small sample distribution under the null hypothesis. This procedure is performed by the <code>granger_test</code> function whose syntax is given below:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a><span class="kw">granger_test</span>(root, shoot, times, <span class="dt">theta =</span> <span class="ot">NULL</span>, <span class="dt">alternative =</span> <span class="st">&quot;twodir&quot;</span>, </span>
<span id="cb8-2"><a href="#cb8-2"></a>             <span class="dt">H =</span> <span class="dv">100</span>, <span class="dt">seed =</span> <span class="dv">123</span>, <span class="dt">showprogress =</span> <span class="ot">TRUE</span>)</span></code></pre></div>
<p>The first two arguments are the first (<code>root</code>) and second (<code>shoot</code>) signals for which we want to test a Granger-Causal relationship. The <code>times</code> argument is the numeric vector containing the time points at which the measurements were taken (in the same unit of time) while <code>theta</code> represents numeric vector (of length 8) containing the starting values for the parameter vector <span class="math inline">\(\theta\)</span> described earlier (if left empty, the starting values will be estimated by the function itself). The <code>alternative</code> argument represents the alternative hypothesis we’re interested in testing: by default we have <code>alternative = &quot;twodir&quot;</code> which indicates that we’re testing the alternative hypothesis:</p>
<p>“<span class="math inline">\(H_A\)</span>: The two signals have a Granger-Causal impact on each other”.</p>
<p>The other possible alternative syntax is the following:</p>
<ol style="list-style-type: decimal">
<li><code>&quot;rtos&quot;</code>: “<span class="math inline">\(H_A\)</span>: the first signal (root) has a Granger-Causal impact on the second signal (shoot)”.</li>
<li><code>&quot;stor&quot;</code>: “<span class="math inline">\(H_A\)</span>: the second signal (shoot) has a Granger-Causal impact on the first signal (root)”.</li>
</ol>
<p>The final arguments to the function concern the parametric bootstrap procedure used to generate the null distribution:</p>
<ul>
<li><code>H</code>: the number of bootstrap replicates. The larger this number is, the better the approximation of the null distribution but the slower the procedure will be to perform the test.</li>
<li><code>seed</code>: the seed value used to generate the signals for the bootstrap distribution. A fixed seed allows to replicate the exact results of the test.</li>
<li><code>showprogress</code>: a boolean parameter allowing to visualize (or not) a progress bar to update the user on the progress of the bootstrap procedure (particularly useful with large values of <code>H</code> and/or longer signals).</li>
</ul>
<p>To perform this parametric bootstrap procedure, the <code>granger_test</code> uses the <code>sim_proc</code> function also made available by the <code>irg</code> package. Let us suppose that we want to simulate bivariate signals from the above mentioned model where only the second signal (shoot) has a Granger-Causal impact on the first and are measured at the same time points as the <code>arabidopsis</code> data also mentioned earlier. Then we could simulate this process as follows:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a>theta_sim &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">1</span>, <span class="fl">0.99</span>, <span class="dv">10</span>, <span class="fl">0.01</span>, <span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="fl">0.1</span>) <span class="co"># the parameter vector used to simulate data</span></span>
<span id="cb9-2"><a href="#cb9-2"></a>times &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">5</span>, <span class="dv">10</span>, <span class="dv">15</span>, <span class="dv">20</span>, <span class="dv">30</span>, <span class="dv">45</span>, <span class="dv">60</span>, <span class="dv">90</span>, <span class="dv">120</span>) <span class="co"># the measurement time points</span></span>
<span id="cb9-3"><a href="#cb9-3"></a></span>
<span id="cb9-4"><a href="#cb9-4"></a><span class="kw">set.seed</span>(<span class="dv">223</span>) <span class="co"># set seed for example replicability</span></span>
<span id="cb9-5"><a href="#cb9-5"></a>sim &lt;-<span class="st"> </span><span class="kw">sim_proc</span>(theta_sim, times) <span class="co"># simulate the bivariate time series</span></span></code></pre></div>
<p>It can be noticed how the values for <span class="math inline">\(\psi_{shoot}\)</span> and <span class="math inline">\(\gamma_{shoot}\)</span> are zero, indicating that the first signal (root) has no impact on the second (shoot). The object <code>sim</code> is a list containing the simulated values for the first signal (<code>$root</code>) and second signal (<code>$shoot</code>) and can now be used within the <code>granger_test</code> function as follows:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a><span class="kw">granger_test</span>(<span class="dt">root =</span> sim<span class="op">$</span>root, <span class="dt">shoot =</span> sim<span class="op">$</span>shoot, <span class="dt">times =</span> times, <span class="dt">alternative =</span> <span class="st">&quot;stor&quot;</span>, <span class="dt">H =</span> <span class="dv">100</span>, <span class="dt">showprogress =</span> <span class="ot">FALSE</span>)</span></code></pre></div>
<pre><code>## $alternative
## [1] &quot;stor&quot;
## 
## $stat
## [1] 31.25822
## 
## $pvalue
## [1] 0.00990099
## 
## $parameters
##   Shoot to Root Intensity Shoot to Root Impact Time 
##                 0.9999624                10.3444944</code></pre>
<p>We can see that the output of this function is a list with the following four elements:</p>
<ol style="list-style-type: decimal">
<li><code>alternative</code>: the code of the alternative hypothesis being tested.</li>
<li><code>stat</code>: the value of the LRT statistic for the Granger-Causal alternative hypothesis being tested.</li>
<li><code>pvalue</code>: the p-value of the test, with a small p-value indicating evidence to reject the null hypothesis in favour of the specified alternative.</li>
<li><code>parameters</code>: the values of the parameters of interest for the specific alternative hypothesis (i.e. the <span class="math inline">\(\psi\)</span> and <span class="math inline">\(\gamma\)</span> parameters).</li>
</ol>
<p>In this simulated example we see that, at a significance level of 0.01, we can reject the null hypothesis in favour of the alternative hypothesis (indeed, in this case we obtain the smallest p-value attainable given the value of <code>H</code>). Hence, having rejected the null hypothesis we can go and interpret the parameter values which, as can be seen, are very close to the true parameter values from which we simulated.</p>
<p>Having seen the performance of the test in a controlled simulated setting, let us apply the test to two signals from the real data contained in <code>arabidopsis</code> and analysed in <span class="citation">Heerah et al. (2021)</span>. This time we (i) specify a starting value for the parameter vector, (ii) use a larger value of <code>H</code> for a better approximation of the null distribution and (iii) test the alternative hypothesis that the first signal (root) has a Granger-Causal impact on the second signal (shoot).</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a>root  &lt;-<span class="st"> </span>arabidopsis<span class="op">$</span>roots[<span class="st">&quot;AT4G27950&quot;</span>, ]</span>
<span id="cb12-2"><a href="#cb12-2"></a>shoot &lt;-<span class="st"> </span>arabidopsis<span class="op">$</span>shoots[<span class="st">&quot;AT4G34750&quot;</span>, ]</span>
<span id="cb12-3"><a href="#cb12-3"></a>theta_start &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">1</span>, <span class="fl">0.5</span>, <span class="dv">10</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="fl">0.5</span>, <span class="dv">10</span>, <span class="dv">1</span>) <span class="co"># starting values for parameter vector</span></span>
<span id="cb12-4"><a href="#cb12-4"></a></span>
<span id="cb12-5"><a href="#cb12-5"></a><span class="kw">granger_test</span>(root, shoot, <span class="dt">times =</span> times, <span class="dt">theta =</span> theta_start, <span class="dt">alternative =</span> <span class="st">&quot;rtos&quot;</span>, <span class="dt">H =</span> <span class="dv">1000</span>, <span class="dt">showprogress =</span> <span class="ot">FALSE</span>)</span></code></pre></div>
<pre><code>## $alternative
## [1] &quot;rtos&quot;
## 
## $stat
## [1] 10.27819
## 
## $pvalue
## [1] 0.000999001
## 
## $parameters
##   Root to Shoot Intensity Root to Shoot Impact Time 
##                 0.9890142                 6.3894862</code></pre>
<p>Also in this case, the p-value is the smallest possible given the value of <code>H</code> and we can reject the null hypothesis in favour of the specified alternative. Having done so, we can see that the second signal (shoot) is positively impacted by the first signal (root) meaning that the second signal increases if the first does so too (and viceversa). In addition, this impact is maximal at a distance in time of roughly 6.5 minutes.</p>
</div>
<div id="references" class="section level2">
<h2>References</h2>
<p>This package is developed based on the work put forward in <span class="citation">Heerah et al. (2021)</span>.</p>
<div id="refs" class="references">
<div id="ref-heerah2021granger">
<p>Heerah, Sachin, Roberto Molinari, Stéphane Guerrier, and Amy Marshall-Colon. 2021. “Granger-Causal Testing for Irregularly Sampled Time Series with Application to Nitrogen Signalling in Arabidopsis.” <em>Bioinformatics</em> 37 (16): 2450–60.</p>
</div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
