<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>`irg` Vignette • irg</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="`irg` Vignette">
<meta property="og:description" content="irg">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">irg</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/vignette.html">Vignette</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/SMAC-Group/irg">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>
<code>irg</code> Vignette</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/SMAC-Group/irg/blob/master/vignettes/vignette.Rmd"><code>vignettes/vignette.Rmd</code></a></small>
      <div class="hidden name"><code>vignette.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">irg</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">mgcv</span>)</pre></body></html></div>
<p>The <code>irg</code> package is based on the procedure proposed in <span class="citation">Heerah et al. (2020)</span>. The functions made available allow to perform Granger-Causal testing for signals that are sampled at the same but irregularly spaced time points. Before using the functions in the package, the signals need to be pre-processed such that they are detrended (mean-stationary) and standardized (normalized). Below we show some examples of possible ways of pre-processing the signals as well as how to use the main functions of the package.</p>
<div id="data-pre-processing" class="section level2">
<h2 class="hasAnchor">
<a href="#data-pre-processing" class="anchor"></a>Data Pre-Processing</h2>
<p>In this section we use data that was collected during an experiment that measured the time-evolved transcriptome of the three replicates of Arabidopsis roots and shoots. The data <code>raw_signals</code> contains a matrix reporting a gene expression count for root and shoot collected at the following time points: 0, 5, 10, 15, 20, 30, 45, 60, 90 and 120 (minutes).</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="no">raw_data</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">raw_data</span>)</pre></body></html></div>
<pre><code>##      root shoot time
## [1,] 4965   674    0
## [2,] 4776   714    0
## [3,] 4806   796    0
## [4,] 4890   540    5
## [5,] 3896   722    5
## [6,] 4441   578    5</code></pre>
<p>We decide to use the following simple functions to pre-process the data in this matrix (but other more appropriate pre-processing can be chosen).</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="co"># Average signals at matching time points and then standardize</span>
<span class="no">avg</span> <span class="kw">&lt;-</span> <span class="kw">function</span>(<span class="no">signals</span>) {

  <span class="no">signal</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="fl">NA</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">signals</span>)/<span class="fl">3</span>)

  <span class="kw">for</span>(<span class="no">i</span> <span class="kw">in</span> <span class="fl">1</span>:(<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">signals</span>)/<span class="fl">3</span>)) {

    <span class="no">signal</span>[<span class="no">i</span>] <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span>(<span class="no">signals</span>[(<span class="fl">3</span>*(<span class="no">i</span>-<span class="fl">1</span>) + <span class="fl">1</span>) : (<span class="fl">3</span>*<span class="no">i</span>)])

  }

  <span class="no">signal_std</span> <span class="kw">&lt;-</span> (<span class="no">signal</span> - <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span>(<span class="no">signal</span>))/<span class="fu"><a href="https://rdrr.io/r/stats/sd.html">sd</a></span>(<span class="no">signal</span>)

  <span class="fu"><a href="https://rdrr.io/r/base/function.html">return</a></span>(<span class="no">signal_std</span>)

}

<span class="co"># Pre-processing function that uses GAM to detrend the signals and then uses previous function to aggregate and normalize</span>
<span class="no">pre_process</span> <span class="kw">&lt;-</span> <span class="kw">function</span>(<span class="no">dat_mat</span>) {

  <span class="no">root</span> <span class="kw">&lt;-</span> <span class="no">dat_mat</span>[,<span class="fl">1</span>]
  <span class="no">shoot</span> <span class="kw">&lt;-</span> <span class="no">dat_mat</span>[,<span class="fl">2</span>]
  <span class="no">time</span> <span class="kw">&lt;-</span> <span class="no">dat_mat</span>[,<span class="fl">3</span>]

  <span class="no">mod_root</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mgcv/man/gam.html">gam</a></span>(<span class="no">root</span> ~ <span class="fu"><a href="https://rdrr.io/pkg/mgcv/man/s.html">s</a></span>(<span class="no">time</span>))
  <span class="no">mod_shoot</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/mgcv/man/gam.html">gam</a></span>(<span class="no">shoot</span> ~ <span class="fu"><a href="https://rdrr.io/pkg/mgcv/man/s.html">s</a></span>(<span class="no">time</span>))

  <span class="no">resid_root</span> <span class="kw">&lt;-</span> <span class="fu">avg</span>(<span class="fu"><a href="https://rdrr.io/r/stats/residuals.html">resid</a></span>(<span class="no">mod_root</span>))
  <span class="no">resid_shoot</span> <span class="kw">&lt;-</span> <span class="fu">avg</span>(<span class="fu"><a href="https://rdrr.io/r/stats/residuals.html">resid</a></span>(<span class="no">mod_shoot</span>))

  <span class="no">out</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">root</span> <span class="kw">=</span> <span class="no">resid_root</span>, <span class="kw">shoot</span> <span class="kw">=</span> <span class="no">resid_shoot</span>)

  <span class="fu"><a href="https://rdrr.io/r/base/function.html">return</a></span>(<span class="no">out</span>)

}</pre></body></html></div>
<p>If we apply the `<code>pre_process</code> function to the raw signals we obtain the following:</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="no">gene_signals</span> <span class="kw">&lt;-</span> <span class="fu">pre_process</span>(<span class="no">raw_data</span>)
<span class="no">gene_signals</span></pre></body></html></div>
<pre><code>## $root
##  [1]  1.71628640 -0.04956446 -1.59010074 -0.59163686 -0.79322148  0.73618577
##  [7]  0.14033187  1.08903416 -0.86743386  0.21011919
## 
## $shoot
##  [1]  1.89019661 -0.71701506 -1.10224276 -0.54846938 -1.11158441  1.36759941
##  [7]  0.39456993  0.05497426 -0.34491931  0.11689071</code></pre>
<p>The result of the pre-processing is saved in the <code>signals</code> data made available with the <code>irg</code> package.</p>
</div>
<div id="granger-causal-testing" class="section level2">
<h2 class="hasAnchor">
<a href="#granger-causal-testing" class="anchor"></a>Granger-Causal Testing</h2>
<p>We can now make use of the <code>irg</code> package functions and for this we make use of the pre-processed signals from the previous section.</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="no">signals</span>)</pre></body></html></div>
<p>The null hypothesis for all tests is the following: “<span class="math inline">\(H_0\)</span>: The two signals do not Granger-cause each other”. Considering this, suppose that we’re interested in computing the test statistic for the alternative hypothesis “<span class="math inline">\(H_A\)</span>: The first signal (root) Granger-causes the second signal (shoot)”, then we would make use of the <code>causal_stat</code> function. In order to use this function, aside from the measurements of the two signals (root and shoot), there are other two main arguments to the function: <code>times</code> and <code>theta</code>. The first one (<code>times</code>) is simply a numeric vector containing the measurement times (with the same unit) and, for this data, these were given in the previous section. The second argument of relevance (<code>theta</code>) is the parameter vector used as the starting values when optimizing the likelihood functions for the test. This vector contains eight values representing the main parameters of interest for the full model (the remaining parameters are implicitly defined in the optimization process). More specifically, the starting values represent the parameters assuming that both signals Granger-cause each other and therefore this vector can be represented as follows:</p>
<p><span class="math display">\[\theta := \big[\phi_{root},\, \psi_{root},\, \gamma_{root},\, \sigma_{root}^2,\, \phi_{shoot},\, \psi_{shoot},\, \gamma_{shoot},\, \sigma_{shoot}^2\big].\]</span></p>
<p>The main parameters of interest are the <span class="math inline">\(\psi\)</span> and <span class="math inline">\(\gamma\)</span> parameters which represent the “intensity” and “time” of (Granger-causal) impact of one signal on the other. More specifically, <span class="math inline">\(\psi \in (-1, 1)\)</span> can be interpreted similarly to Pearson’s correlation coefficient: the closer the value is to 1 (in absolute value) the greater is the impact of one signal on the other, while the sign represents the direction of the impact where, for example, a negative value implies that the impacted signal grows as the “causing” signal descreases (and viceversa). The <span class="math inline">\(\gamma \in \mathbb{R}^+\)</span> parameter instead represents the distance in time at which the impact of one signal on the other is maximal (under the assumption that the impact increases and then decreases monotonically over time). So, for example, <span class="math inline">\(\psi_{root}\)</span> represents the intensity of the impact of the shoot (second signal) on the root (first signal) while <span class="math inline">\(\gamma_{root}\)</span> represents the time required for the impact of the shoot on the root to be maximal.</p>
<p>While defining the <code>times</code> argument is straightforward, specifying the starting values for the parameter vector is not necessarily easy. Given that the signals are standardized, a generally acceptable starting value could be <code><a href="https://rdrr.io/r/base/c.html">c(1, 0.5, 1, 1, 1, 0.5, 1, 1)</a></code>, however the choice of the <span class="math inline">\(\gamma\)</span> parameters should be in the range of the distances between adjacent time points (for the considered data, a value of <span class="math inline">\(\gamma = 10\)</span> could be reasonable). Finally, in order to estimate these parameters within their admissable ranges, these values need to be transformed through the natural logarithm (<code>log</code>) while the <span class="math inline">\(\psi\)</span> parameters are transformed and scaled within the <code>logit2</code> function included in the <code>irg</code> package (see code below for an example).</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="no">times</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0</span>, <span class="fl">5</span>, <span class="fl">10</span>, <span class="fl">15</span>, <span class="fl">20</span>, <span class="fl">30</span>, <span class="fl">45</span>, <span class="fl">60</span>, <span class="fl">90</span>, <span class="fl">120</span>) <span class="co"># vector of measurement times</span>
<span class="no">theta</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span>(<span class="fl">1</span>), <span class="fu"><a href="../reference/logit2.html">logit2</a></span>(<span class="fl">0.5</span>*<span class="fl">2</span>), <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span>(<span class="fl">10</span>), <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span>(<span class="fl">1</span>), <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span>(<span class="fl">1</span>), <span class="fu"><a href="../reference/logit2.html">logit2</a></span>(<span class="fl">0.5</span>*<span class="fl">2</span>), <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span>(<span class="fl">10</span>), <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span>(<span class="fl">1</span>)) <span class="co"># transformed starting values</span></pre></body></html></div>
<p>Now that all arguments to the function are defined, these can be used within the function to obtain the Granger-causal test statistic for a given alternative hypothesis of interest. For this reason, the last argument to the function is <code>alternative</code> which allows to specify the desired alternative hypothesis and the options are</p>
<ol style="list-style-type: decimal">
<li>
<code>"rtos"</code>: “<span class="math inline">\(H_A\)</span>: the first signal (root) Granger-causes the second signal (shoot)”.</li>
<li>
<code>"stor"</code>: “<span class="math inline">\(H_A\)</span>: the second signal (shoot) Granger-causes the first signal (root)”.</li>
<li>
<code>"twodir"</code>: “<span class="math inline">\(H_A\)</span>: both signals Granger-cause each other”.</li>
</ol>
<p>In this example we will test the alternative “<span class="math inline">\(H_A\)</span>: the first signal (root) Granger-causes the second signal (shoot)” and therefore specify <code>alternative = "rtos"</code>.</p>
<p>We can see that the output of the function is a list in which the first element is the test statistic (the closer to zero, the less evidence in favour of the alternative hypothesis), while the second element contains the parameters of interest for the alternative hypothesis which, in this case, are the estimated values of <span class="math inline">\(\psi_{shoot}\)</span> and <span class="math inline">\(\gamma_{shoot}\)</span>. In order to test whether we can reject the null hypothesis in favour of an alternative hypothesis, we need to obtain the distribution of the test statistic under the null hypothesis.</p>
<p>The main function of the <code>irg</code> package is the <code>granger_test</code> function which makes use of the <code>causal_stat</code> function to obtain a bootstrap distribution of the test statistic under the null hypothesis and then computes a p-value to be compared to a desired level of significance <span class="math inline">\(\alpha\)</span>. The arguments to this function are the same as for the <code>causal_stat</code> function with the following exceptions:</p>
<ul>
<li>By default the <code>theta</code> argument is set to <code>NULL</code> which, if left this way, uses an approximate method to estimate reasonable starting values for the full parameter vector.</li>
<li>The <code>granger_test</code> function has two additional arguments:
<ul>
<li>
<code>H</code>: the number of boostrap replicates of the test statistic under the null hypothesis. The larger this number, the better the approximation of the bootstrap distribution to the true null distribution. By defult the value is 100 which allows the p-value to be computed in a reasonable amount of time (depending on the sample sizes) while larger values will obviously require higher computational times (although obtaining better approximations).</li>
<li>
<code>seed</code>: this is a value that allows to control the random generation of the null distribution, implying that the results on specific signals will remain the same if the seed is fixed (by default <code>seed = 123</code>).</li>
</ul>
</li>
</ul>
<p>Let us understand if the previous test statistic allows us to reject the null hypothesis in favour of the specified alternative.</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="fu"><a href="../reference/granger_test.html">granger_test</a></span>(<span class="kw">root</span> <span class="kw">=</span> <span class="no">gene_signals</span>$<span class="no">root</span>, <span class="kw">shoot</span> <span class="kw">=</span> <span class="no">gene_signals</span>$<span class="no">shoot</span>, <span class="kw">times</span> <span class="kw">=</span> <span class="no">times</span>, <span class="kw">alternative</span> <span class="kw">=</span> <span class="st">"rtos"</span>, <span class="kw">H</span> <span class="kw">=</span> <span class="fl">10</span>)</pre></body></html></div>
<pre><code>## 
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |=======                                                               |  10%
  |                                                                            
  |==============                                                        |  20%
  |                                                                            
  |=====================                                                 |  30%
  |                                                                            
  |============================                                          |  40%
  |                                                                            
  |===================================                                   |  50%
  |                                                                            
  |==========================================                            |  60%
  |                                                                            
  |=================================================                     |  70%
  |                                                                            
  |========================================================              |  80%
  |                                                                            
  |===============================================================       |  90%
  |                                                                            
  |======================================================================| 100%</code></pre>
<pre><code>## $alternative
## [1] "rtos"
## 
## $pvalue
## [1] 0.7272727
## 
## $parameters
##   Root to Shoot Intensity Root to Shoot Impact Time 
##                 0.7139067                17.5221842</code></pre>
</div>
<div id="references" class="section level2">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<p>This package is developed based on the work put forward in <span class="citation">Heerah et al. (2020)</span>.</p>
<div id="refs" class="references">
<div id="ref-Heerah2020">
<p>Heerah, Sachin, Roberto Molinari, Stéphane Guerrier, and Amy Marshall-Colon. 2020. “Granger-Causal Testing for Irregularly Sampled Time Series with Application to Nitrogen Signaling in Arabidopsis.” <em>bioRxiv</em>. Cold Spring Harbor Laboratory. <a href="https://doi.org/10.1101/2020.06.15.152819">https://doi.org/10.1101/2020.06.15.152819</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Roberto Molinari, Stephane Guerrier.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
